src/ordered_dict.jl

```diff
@@ -24,19 +24,34 @@ type OrderedDict{K,V} <: Associative{K,V}
     function OrderedDict()
         new(zeros(Int32,16), Array(K,0), Array(V,0), 0, false)
     end
+    function OrderedDict(n::Integer)
+        slotsz = max(16, (n*3)>>1 )
+        new(zeros(Int32,slotsz), Array(K,n), Array(V,n), 0, false)
+    end
     function OrderedDict(kv)
-        h = OrderedDict{K,V}()
+        h = OrderedDict{K,V}(length(kv))
```

Comment:
Thanks for having a look. I assumed that there had to be a 'length' method
for 'kv' because it was used in a for loop

for (k,v) in kv

, but I guess 'for' only requires the three functions necessary for an
iterator.  Do I want/need to do some kind of Traits thing to dispatch on
whether or not 'kv' has 'length' and then fall back to the old version if
it does not?

While I've got you on the line, do you have any thoughts about having an
inner constructor that makes an empty OrderedDict that can only be used in
the context of these filling functions? Perhaps it would be better to
copy/paste

slotsz = max(16, (n*3)>>1 )
h = new(zeros(Int32,slotsz), Array(K,n), Array(V,n), 0, false)

into the two relevant places.


Pete

____________________
Peter M. Haverty, Ph.D.
Genentech, Inc.
phaverty@gene.com

On Wed, Jul 27, 2016 at 2:45 PM, Keno Fischer <notifications@github.com>
wrote:

> In src/ordered_dict.jl
> <https://github.com/JuliaLang/DataStructures.jl/pull/211#discussion_r72530227>
> :
>
> >      function OrderedDict(kv)
> > -        h = OrderedDict{K,V}()
> > +        h = OrderedDict{K,V}(length(kv))
>
> This introduces an assumption that kv supports length.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/DataStructures.jl/pull/211/files/6e26b3fd7faa3ad1303fec37e180b3adcfe9171c#r72530227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH02K9RnHMKNrLSC7RY_I4MrxsB7uY6Uks5qZ9GPgaJpZM4JVhGj>
> .
>